src/rgw/rgw_op.cc

```diff
@@ -2362,16 +2381,53 @@ void RGWGetObj::execute(optional_yield y)
     } catch (const buffer::error&) {}
   }
 
-
-  if (get_type() == RGW_OP_GET_OBJ && get_data) {
-    op_ret = handle_cloudtier_obj(attrs, sync_cloudtiered);
-    if (op_ret < 0) {
-      ldpp_dout(this, 4) << "Cannot get cloud tiered object: " << *s->object
-          <<". Failing with " << op_ret << dendl;
-      if (op_ret == -ERR_INVALID_OBJECT_STATE) {
-        s->err.message = "This object was transitioned to cloud-s3";
+  attr_iter = attrs.find(RGW_ATTR_MANIFEST);
+  if (attr_iter != attrs.end()) {
+    RGWObjManifest m;
+    try {
+    decode(m, attr_iter->second);
+    if (get_type() == RGW_OP_GET_OBJ && get_data && m.get_tier_type() == "cloud-s3") {
+      RGWObjTier tier_config;
+      m.get_tier_config(&tier_config);
+      if (tier_config.tier_placement.allow_read_through) {
+        uint64_t days = tier_config.tier_placement.read_through_restore_days;
+        attr_iter = attrs.find(RGW_ATTR_RESTORE_STATUS);
+        rgw::sal::RGWRestoreStatus restore_status;
+        if (attr_iter != attrs.end()) {
+          bufferlist bl = attr_iter->second;
+          auto iter = bl.cbegin();
+          decode(restore_status, iter);
+        }
+        if (attr_iter == attrs.end() || restore_status == rgw::sal::RGWRestoreStatus::RestoreFailed) {
+          op_ret = handle_cloudtier_obj(s, this, driver, days, attrs, sync_cloudtiered, tier_config, y);
+          if (op_ret < 0) {
+            ldpp_dout(this, 4) << "Cannot get cloud tiered object: " << *s->object
+            <<". Failing with " << op_ret << dendl;
+            if (op_ret == -ERR_INVALID_OBJECT_STATE) {
+              s->err.message = "failed to restore object";
+            }
+          goto done_err;
+          }
+        }
+        if (attr_iter == attrs.end() || restore_status != rgw::sal::RGWRestoreStatus::CloudRestored) {
+          /* * if the attr is not found, we assume restore is not complete
+           * So the first read through request will return but actually downloaded
+           * object asyncronously.
+           */
+          op_ret = -ERR_INVALID_OBJECT_STATE;
+          ldpp_dout(this, 5) << "restore is not completed yet, please retry again" << dendl;
+          goto done_err;
+          s->err.message = "restore is not completed yet";
+        }
+      } else {
+        op_ret == -ERR_INVALID_OBJECT_STATE;
+        s->err.message = "Read through is not enabled for this config";
```

Comment:
Could you add earlier message too - ""This object was transitioned to cloud-s3".. probably
""This object was transitioned to cloud-s3 and Read-through is not enabled".